src/bin/bobp.rs

```diff
@@ -1,150 +1,494 @@
+) {
     let pause = time::Duration::from_millis(period_ms);
     let mut last_put_count = stat.put_total.load(Ordering::Relaxed);
     let mut last_get_count = stat.get_total.load(Ordering::Relaxed);
-
+    let mut last_st_put_time = stat.put_time_ns_single_thread.load(Ordering::Relaxed);
+    let mut last_st_put_count = stat.put_count_single_thread.load(Ordering::Relaxed);
+    let mut last_st_get_time = stat.get_time_ns_single_thread.load(Ordering::Relaxed);
+    let mut last_st_get_count = stat.get_count_single_thread.load(Ordering::Relaxed);
+    let k = request_bytes as f64 / period_ms as f64 * 1000.0 / 1024.0;
+    let start = Instant::now();
     while !stop_token.load(Ordering::Relaxed) {
         thread::sleep(pause);
-        let cur_put_count = stat.put_total.load(Ordering::Relaxed);
-        let put_count_spd = (cur_put_count - last_put_count) * 1000 / period_ms;
-        last_put_count = cur_put_count;
-
-        let cur_get_count = stat.get_total.load(Ordering::Relaxed);
-        let get_count_spd = (cur_get_count - last_get_count) * 1000 / period_ms;
-        last_get_count = cur_get_count;
+        let put_count_spd = get_diff(&mut last_put_count, stat.put_total.load(Ordering::Relaxed))
+            * 1000
+            / period_ms;
+        let get_count_spd = get_diff(&mut last_get_count, stat.get_total.load(Ordering::Relaxed))
+            * 1000
+            / period_ms;
+        let cur_st_put_time = get_diff(
+            &mut last_st_put_time,
+            stat.put_time_ns_single_thread.load(Ordering::Relaxed),
+        ) as f64;
+        let cur_st_put_count = get_diff(
+            &mut last_st_put_count,
+            stat.put_count_single_thread.load(Ordering::Relaxed),
+        ) as f64;
+        let cur_st_get_time = get_diff(
+            &mut last_st_get_time,
+            stat.get_time_ns_single_thread.load(Ordering::Relaxed),
+        ) as f64;
+        let cur_st_get_count = get_diff(
+            &mut last_st_get_count,
+            stat.get_count_single_thread.load(Ordering::Relaxed),
+        ) as f64;
 
         let put_error = stat.put_error.load(Ordering::Relaxed);
         let get_error = stat.get_error.load(Ordering::Relaxed);
+
         println!(
-            "put: {:5} rps | get {:5} rps | put err: {:5} | get err: {:5}",
-            put_count_spd, get_count_spd, put_error, get_error
+            "put: {:>6} rps  | get {:>6} rps   | put err: {:5}     | get err: {:5}\r\n\
+            put: {:>6.2} kb/s | get: {:>6.2} kb/s | put lat: {:>6.2} ms | get lat: {:>6.2} ms",
+            put_count_spd,
+            get_count_spd,
+            put_error,
+            get_error,
+            put_count_spd as f64 * k,
+            get_count_spd as f64 * k,
+            finite_or_default(cur_st_put_time / cur_st_put_count / 1e9),
+            finite_or_default(cur_st_get_time / cur_st_get_count / 1e9)
         );
     }
+    let elapsed = start.elapsed();
+    let elapsed_secs = elapsed.as_secs() as f64;
```

Comment:
better to use inner float representation of the duration [as_secs_f64()](https://doc.rust-lang.org/std/time/struct.Duration.html#method.as_secs_f64)